### ISSUE TITLE
Gradient flow despite `lax.stop_gradient`

### ISSUE THREAD

**Issue by maxecharles** [LouisDesdoigts/amigo#82]
I tried doing a fit with only calibrator exposures --- no flats or validators --- with `FF` and `non_linearity` still in the optimisers. I expected their parameter curves to just be flat because of the `nuke_pixel_grads` call in the simulate method

https://github.com/LouisDesdoigts/amigo/blob/472c0f751bd10d7a59272b3f8c16fdaa7839a5ca/src/amigo/model_fits.py#L389

However I still did observe gradient flow to both parameters. I haven't tested this on `main` branch amigo version though, but I assume it will be the same. 

my best frenemy Claude says:
> nuke_pixel_grads is a no-op for FF and non_linearity when they are present in model_params.
The function applies lax.stop_gradient to these values inside the loss computation, which would correctly block gradient flow if they were part of the static model argument. However, because they are registered in model_params, eqx.filter_value_and_grad treats them as differentiable inputs from the start — it differentiates with respect to the first argument of _val_grad_fn, which is model_params. 

Which checks out, I guess.

I have a fix which works and makes sense with Claude's diagnosis (I will put in retrain branch unless you can think of a better idea). It's just to set the gradients to zero in the training loop rather than trying to nuke them, which it seems like you've already been doing with the `validator_params`.

```python
    def train(
        ...
    ):

        ...
        
        for epoch in looper:

            ...

            for i in batch_inds[epoch]:

                ...

                # Nuke the non-validator grads since we take gradients wrt all parameters
                if "val" in batch_key:
                    grad_params = grads.params
                    for param, value in grad_params.items():
                        if param not in validator_params:
                            if isinstance(value, dict):
                                grad_params[param] = jtu.map(lambda x: x * 0, value)
                            else:
                                grad_params[param] = value * 0
                    grads = grads.set("params", grad_params)
                    
                ######################## THIS BIT IS THE NEW BIT #########################
                if "cal" in batch_key:
                    grad_params = grads.params
                    for param, value in grad_params.items():
                        if param in ["FF", "non_linearity"]:
                            if isinstance(value, dict):
                                grad_params[param] = jtu.map(lambda x: x * 0, value)
                            else:
                                grad_params[param] = value * 0
                    grads = grads.set("params", grad_params)
                ##########################################################################       

                ...
```

Bit confused tbh why this is happening when I clearly have no problem with `lax.stop_gradient` for stopping the aberrations for non-calibrators. I think the difference is the `lax.stop_gradient` is applied to the `coefficients` in the model_params dictionary rather than to a leaf of the model PyTree directly.?

**Comment by LouisDesdoigts** [comment]
98% sure Claude is being an idiot:

> The function applies lax.stop_gradient to these values inside the loss computation, which would correctly block gradient flow if they were part of the static model argument. However, because they are registered in model_params, eqx.filter_value_and_grad treats them as differentiable inputs from the start 

If it was part of the static model arguments we wouldn't be calculating gradients of those parameters in the first place, so this is saying "we wouldnt get gradients for them if we didnt take gradients of them"... Thanks genius :D 

Yeah I really dont like this 'fix' - it doesn't solve the problem it works around it by making the code more bespoke. If that were the case then the lax.stop_grad call wouldn't work on the aberrations... We're better than the clankers. 

Maybe try just manually using `lax.stop_grad` in the loss or printing to make sure its running. There's clearly something _else_ going on.

Also I'm not sure this issue would happen on main.... Otherwise we would see PSF-like structure in the FF and NL maps, since there us MUCH more signal in that data than the flats.

**Comment by maxecharles** [comment]
Yeah idk what was going in. Tested main and the bug didn't show up, and the bug doesn't show up when using the `Trainer` class (as opposed to `ValBatchedTrainer`) on the retrain branch too. So, very confused and will investigate more

**Comment by maxecharles** [comment]
Clanker helped me find the actual issue – gradients from the likelihood were correctly nuked but the FF_reg and NL_reg priors were giving gradients from the calibrator exposures too! So just need to add a `nuke_pixel_grads` call in the regularisation functions as well
